asllib/TopoSort.mli

```diff
@@ -0,0 +1,81 @@
+(** This modules implement a topological sorting algorithm. *)
+
+(** {1 Topological sorting of ASTs.} *)
+
+(** In the following, {i dependencies} are understood in the ASL declaration
+    level. For example, in the following ASL code:
+    {v
+       constant x: integer = 4;
+       constant y: integer = x + 1;
+    v}
+    the declaration of [y] should happen after the declaration of [x],
+    because the declaration of [y] uses the value of [x].
+*)
+
+(** Entry-point for dependency-ordered iterations on ASTs. *)
+module ASTFold : sig
+  type t
+  (** Represents a step in the folding *)
+
+  val build :
+    def:(AST.decl -> AST.identifier) ->
+    extra_def:(AST.decl -> AST.identifier list) ->
+    use:(AST.decl -> AST.identifier list) ->
+    AST.t ->
+    t
+  (** [build ~def ~extra_def ~use ast] is the internal representation of the
```

Comment:
```suggestion
  (** [build ~def ~extra_def ~use ast] returns the internal representation of the
```

Reply:
Moreover, if I am not wrong, all usages of this module immediately apply the `fold` function to the result of `build`. As a consequence it may not be necessary to export the intermediate type `t`. Accordingly `build` would not be exported either and the new type of fold would be something like:
```OCaml
val fold : (step -> 'acc -> 'acc) ->  AST.t -> 'acc -> 'acc
```
Assuming that the `def`, `extra_def` and `use` functions are the same for all usages (which seems to be case), they can be found from the modeule `ASTUtils`.